pennylane_lightning/src/rework/StateVector.cpp

```diff
@@ -0,0 +1,30 @@
+// Copyright 2021 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#include "StateVector.hpp"
+
+Pennylane::StateVector Pennylane::StateVector::create(const pybind11::array_t<CplxType>* numpyArray) {
+    pybind11::buffer_info numpyArrayInfo = numpyArray->request();
+
+    if (numpyArrayInfo.ndim != 1)
+        throw std::invalid_argument("NumPy array must be a 1-dimensional array");
+    if (numpyArrayInfo.itemsize != sizeof(CplxType))
```

Comment:
> double "should" be an IEEE-754 64-bit floating point type (though its not explicitly guaranteed by C++ standards)

Yeah, was curious because of that. 🤔 Agreed, it would more so be implementation dependent. Would it be worth hardcoding the value here?

Reply:
Hardcoding the value would be inadvisable--at the end of the day, `itemsize` must conform to the size of whatever complex type is in use, otherwise the internal numpy array will be misinterpreted when casting to `CplxType*`.

AFAIK, there aren't any standard fixed-width floating point representations (unlike [integers](http://www.cplusplus.com/reference/cstdint/)). However, there does exists a way of checking the floating-point representation is IEEE-754 compliant (e.g. [this](https://stackoverflow.com/questions/5777484/how-to-check-if-c-compiler-uses-ieee-754-floating-point-standard)). 

At the end of the day however, I'm not sure its wise to complicate the implementation by trying to support systems in which `double` isn't IEEE-754 64-bit floating-point compliant. But I'll leave that up to you :)